clients/git/client_test.go

```diff
@@ -0,0 +1,242 @@
+// Copyright 2023 OpenSSF Scorecard Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package git
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	gitV5 "github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	"github.com/ossf/scorecard/v4/clients"
+)
+
+func createTestRepo(t *testing.T) (path string) {
+	t.Helper()
+	dir, err := os.MkdirTemp("", "testrepo")
+	if err != nil {
+		t.Fatalf("Failed to create temporary directory: %v", err)
+	}
+
+	r, err := gitV5.PlainInit(dir, false)
+	if err != nil {
+		t.Fatalf("Failed to initialize git repo: %v", err)
+	}
+
+	w, err := r.Worktree()
+	if err != nil {
+		t.Fatalf("Failed to get worktree: %v", err)
+	}
+
+	// Create a new file
+	filePath := filepath.Join(dir, "file")
+	err = os.WriteFile(filePath, []byte("Hello, World!"), 0o644) //nolint:gosec
+	if err != nil {
+		t.Fatalf("Failed to write a file: %v", err)
+	}
+
+	// Add it to the staging area
+	_, err = w.Add("file")
+	if err != nil {
+		t.Fatalf("Failed to add a file to staging area: %v", err)
+	}
+
+	// Commit
+	_, err = w.Commit("Initial commit", &gitV5.CommitOptions{
+		Author: &object.Signature{
+			Name:  "Test Author",
+			Email: "author@example.com",
+			When:  time.Now(),
+		},
+	})
+
+	if err != nil {
+		t.Fatalf("Failed to make initial commit: %v", err)
+	}
+
+	t.Cleanup(func() {
+		os.RemoveAll(dir) // nolint:errcheck
+	})
+	return dir
+}
+
+//nolint:testparallel
+func TestInitRepo(t *testing.T) {
+	tests := []struct { //nolint:govet
+		name        string
+		uri         string
+		commitSHA   string
+		commitDepth int
+		expectedErr string
+	}{
+		{
+			name:        "Success",
+			uri:         "file://%s",
+			commitSHA:   "HEAD",
+			commitDepth: 1,
+		},
+		{
+			name:        "InvalidUri",
+			uri:         ":",
+			commitSHA:   "",
+			commitDepth: 1,
+			expectedErr: "repository does not exist",
+		},
+		{
+			name:        "InvalidCommitSha",
+			uri:         "file://%s",
+			commitSHA:   "invalid",
+			commitDepth: 1,
+		},
```

Comment:
the fact that this didn't error might be a bug. It looks like `plumbing.NewHash` uses an empty hash on error.
https://github.com/ossf/scorecard/blob/9e719948b37672d569c15370a6d35cf574eb1e37/clients/git/client.go#L90-L99

So if we want to fail on invalid commit sha, we would need to check for hexadecimal ourselves.

Reply:
I am going to remove this test case and create an issue.

Reply:
https://github.com/ossf/scorecard/issues/3392